### ISSUE TITLE
XLA Padding Causing OOM

### ISSUE THREAD

**Issue by nikhilanayak** [google/flax#1733]
Provide as much information as possible. At least, this should include a description of your issue and steps to reproduce the problem. If possible also provide a summary of what steps or workarounds you have already tried.

### Problem you have encountered:
Flax does 32x and 128x expansion, causing the TPUs to run out of memory.

### What you expected to happen:
Flax should be able to successfully allocate this memory, like how it does on CPUs and GPUs.

### Logs, error messages, etc:
```
Total hbm usage >= 11.47G:
    reserved        530.00M 
    program          10.96G 
    arguments            0B 

Output size 0B; shares 0B with arguments.

Program hbm requirement 10.96G:
    global             4.0K
    HLO temp         10.96G (1.2% utilization: Unpadded (140.25M) Padded (10.96G), 0.0% fragmentation (0B))

  Largest program allocations in hbm:

  1. Size: 8.77G
     Operator: op_type="conv_general_dilated" op_name="jit(conv_general_dilated)/conv_general_dilated[\n  batch_group_count=1\n  dimension_numbers=ConvDimensionNumbers(lhs_spec=(0, 4, 1, 2, 3), rhs_spec=(4, 3, 0, 1, 2), out_spec=(0, 4, 1, 2, 3))\n  feature_group_count=1\n  lhs_dilation=(1, 1, 1)\n  lhs_shape=(1, 8, 2176, 1056, 2)\n  padding=((1, 1), (1, 1), (0, 0))\n  precision=None\n  preferred_element_type=None\n  rhs_dilation=(1, 1, 1)\n  rhs_shape=(3, 3, 1, 2, 1)\n  window_strides=(1, 1, 1)\n]" source_file="/home/nikhil/.local/lib/python3.8/site-packages/flax/linen/linear.py" source_line=282
     Shape: f32[8,2176,8,132,1]{4,2,3,1,0:T(8,128)}
     Unpadded size: 70.12M
     Extra memory due to padding: 8.70G (128.0x expansion)
     XLA label: %fusion = f32[8,2176,8,132,1]{4,2,3,1,0:T(8,128)} fusion(bf16[8,2176,8,132,2]{2,4,3,1,0:T(4,128)(2,1)} %copy.10, bf16[3,3,1,2,1]{4,3,2,1,0:T(4,128)(2,1)} %copy.11), kind=kOutput, calls=%fused_computation, metadata={op_type="conv_general_dilated" op_name="j...
     Allocation type: HLO temp
     ==========================

  2. Size: 2.19G
     Shape: bf16[8,2176,8,132,2]{2,4,3,1,0:T(4,128)(2,1)}
     Unpadded size: 70.12M
     Extra memory due to padding: 2.12G (32.0x expansion)
     XLA label: %copy.10 = bf16[8,2176,8,132,2]{2,4,3,1,0:T(4,128)(2,1)} copy(f32[8,2176,8,132,2]{1,4,3,2,0:T(2,128)} %bitcast)
     Allocation type: HLO temp
     ==========================

  3. Size: 4.0K
     Shape: u32[8,128]{1,0}
     Unpadded size: 4.0K
     XLA label: constant literal
     Allocation type: global
     ==========================

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

```
```
RuntimeError: RESOURCE_EXHAUSTED: Ran out of memory in memory space hbm. Used 10.96G of 7.48G hbm. Exceeded hbm capacity by 3.47G.

Total hbm usage >= 11.47G:
    reserved        530.00M 
    program          10.96G 
    arguments            0B 

Output size 0B; shares 0B with arguments.

Program hbm requirement 10.96G:
    global             4.0K
    HLO temp         10.96G (1.2% utilization: Unpadded (140.25M) Padded (10.96G), 0.0% fragmentation (0B))

  Largest program allocations in hbm:

  1. Size: 8.77G
     Operator: op_type="conv_general_dilated" op_name="jit(conv_general_dilated)/conv_general_dilated[\n  batch_group_count=1\n  dimension_numbers=ConvDimensionNumbers(lhs_spec=(0, 4, 1, 2, 3), rhs_spec=(4, 3, 0, 1, 2), out_spec=(0, 4, 1, 2, 3))\n  feature_group_count=1\n  lhs_dilation=(1, 1, 1)\n  lhs_shape=(1, 8, 2176, 1056, 2)\n  padding=((1, 1), (1, 1), (0, 0))\n  precision=None\n  preferred_element_type=None\n  rhs_dilation=(1, 1, 1)\n  rhs_shape=(3, 3, 1, 2, 1)\n  window_strides=(1, 1, 1)\n]" source_file="/home/nikhil/.local/lib/python3.8/site-packages/flax/linen/linear.py" source_line=282
     Shape: f32[8,2176,8,132,1]{4,2,3,1,0:T(8,128)}
     Unpadded size: 70.12M
     Extra memory due to padding: 8.70G (128.0x expansion)
     XLA label: %fusion = f32[8,2176,8,132,1]{4,2,3,1,0:T(8,128)} fusion(bf16[8,2176,8,132,2]{2,4,3,1,0:T(4,128)(2,1)} %copy.10, bf16[3,3,1,2,1]{4,3,2,1,0:T(4,128)(2,1)} %copy.11), kind=kOutput, calls=%fused_computation, metadata={op_type="conv_general_dilated" op_name="j...
     Allocation type: HLO temp
     ==========================

  2. Size: 2.19G
     Shape: bf16[8,2176,8,132,2]{2,4,3,1,0:T(4,128)(2,1)}
     Unpadded size: 70.12M
     Extra memory due to padding: 2.12G (32.0x expansion)
     XLA label: %copy.10 = bf16[8,2176,8,132,2]{2,4,3,1,0:T(4,128)(2,1)} copy(f32[8,2176,8,132,2]{1,4,3,2,0:T(2,128)} %bitcast)
     Allocation type: HLO temp
     ==========================

  3. Size: 4.0K
     Shape: u32[8,128]{1,0}
     Unpadded size: 4.0K
     XLA label: constant literal
     Allocation type: global
     ==========================
```

### Steps to reproduce:
I'm having trouble getting FLAX to work on Colab. If I get that working, I'll create a colab notebook.

**Comment by marcvanzee** [comment]
Hi, could you add a code snippet? Doesn't have to be Colab, you can just put it in a codeblock here and I will run it mysef.

**Comment by nikhilanayak** [comment]
Here's some code that should recreate the same issue:
```py
import jax
from tqdm import tqdm
from jax import jit
import numpy as np
import jax.numpy as jnp
from flax import linen as nn
from flax.training import train_state
from flax import optim
import optax
from flax.training import train_state

from tensorflow.keras.utils import Sequence
class Dataset(Sequence):
    def __init__(self, batch_size):
        self.batch_size = batch_size

    def __len__(self):
        return 1000

    def __getitem__(self, i):
        return np.zeros((8, 2048, 1952, 2, 1))


rng = jax.random.PRNGKey(0)
rng, key = jax.random.split(rng)


def conv_and_pool(layer, features, kernel_size):
	x = nn.Conv(features=features, kernel_size=kernel_size)(layer)
	x = nn.relu(x)
	x = nn.max_pool(x, window_shape=(1, 2, 2), strides=(1, 2, 2), padding="SAME")
	#print(x.shape)
	return x

def convtrans_layer(layer, features, kernel_size):
  x = nn.ConvTranspose(features, kernel_size=kernel_size, strides=(1, 2, 2), padding="SAME")(layer)
  x = nn.relu(x)
  return x

class Encoder(nn.Module):
	@nn.compact
	def __call__(self, input_layer):
		x = conv_and_pool(input_layer, features=2, kernel_size=(3, 3, 1))
		x = conv_and_pool(x, features=4, kernel_size=(3, 3, 1))
		x = conv_and_pool(x, features=8, kernel_size=(3, 3, 1))
		x = conv_and_pool(x, features=16, kernel_size=(3, 3, 1))
		x = conv_and_pool(x, features=32, kernel_size=(3, 3, 1))

		return x
  
class Decoder(nn.Module):

	@nn.compact
	def __call__(self, input_layer):
		x = convtrans_layer(input_layer, 32, kernel_size=(3, 3, 1))
		x = convtrans_layer(x, 16, kernel_size=(3, 3, 1))
		x = convtrans_layer(x, 8, kernel_size=(3, 3, 1))
		x = convtrans_layer(x, 4, kernel_size=(3, 3, 1))
		x = convtrans_layer(x, 2, kernel_size=(3, 3, 1))
		x = nn.Conv(1, (3, 3, 1), padding="SAME")(x)

		return x

class Autoencoder(nn.Module):
	def setup(self):
		self.encoder = Encoder()
		self.decoder = Decoder()

	def __call__(self, x):
		out = self.encoder(x)
		reconstructed = self.decoder(out)
		return reconstructed, x 
	
	def generate(self, x):
		params = self.get_param("decoder")
		return Decoder.call(params, x)


def compute_metrics(generated, real):
	error = generated - real
	loss = jnp.mean(jnp.square(error))
	metrics = {
		"loss": loss
	}		
	return metrics


dataset = Dataset(8)

def train_step(state):
	def loss_fn(params):
		data = dataset[0]

		rec, out = Autoencoder().apply({"params": params}, data)
		loss = compute_metrics(rec, out)["loss"]
		print(loss)
		return jnp.array(loss)

	grads = jax.grad(loss_fn)(state.params)
	return state.apply_gradients(grads=grads)
  
print(f"{jax.device_count()} devices")


optimizer = optim.sgd.GradientDescent(learning_rate=0.01)
EPOCHS = 10

state = train_state.TrainState.create(
	apply_fn=Autoencoder().apply,
	params=Autoencoder().init(rng, dataset[0])["params"],
	tx=optax.sgd(0.01)
)

for epoch in range(EPOCHS):
	bar = tqdm(range(len(dataset)))
	for _ in bar:
		state = train_step(state)
		bar.set_description(f"Epoch {epoch}")
```
I'm running this on the TPU v2-8. I also got a similar issue using Tensorflow XLA (it didn't OOM, but it was really slow and was mallocing a lot of memory). 

I also read this from [A Google Cloud Performance Guide](https://cloud.google.com/tpu/docs/performance-guide):

> Arrays in the Cloud TPU are tiled. This entails padding one of the dimensions to a multiple of 8, and a different dimension to a multiple of 128. XLA performs data layout transformations and data is arranged in memory such that the hardware can efficiently process the data. These transformations are driven by heuristics. While they are beneficial in most cases, there is always potential for the compiler to do the wrong thing. To achieve the highest performance, it can be beneficial to experiment with different model configurations.

My data has a shape of (8, 2048, 1952, 2, 1), so this shouldn't be an issue (8 is the batch size, (2048, 1952, 2) is the actual shape).

**Comment by marcvanzee** [comment]
I have a few clarifying questions:

- It seems you aren't `jit`ing your code, is this intended? You should `jit` it for the code to compile on TPU. However, your stack trace does seem to compile for TPU so I guess your code snippet is not correct?
- Why are you expecting this to work? Did it work on GPU, if so, on which GPU?

**Comment by nikhilanayak** [comment]
Hmm, the code I sent doesn't seem to be working.
Try this out, this works on my 3070 in the tensorflow-gpu docker container:
```py
import jax
from tqdm import tqdm
from jax import jit
import numpy as np
import jax.numpy as jnp
from flax import linen as nn
from flax.training import train_state
from flax import optim
import optax
from flax.training import train_state


rng = jax.random.PRNGKey(0)
rng, key = jax.random.split(rng)


def conv_and_pool(layer, features, kernel_size):
	x = nn.Conv(features=features, kernel_size=kernel_size)(layer)
	x = nn.relu(x)
	x = nn.max_pool(x, window_shape=(1, 2, 2), strides=(1, 2, 2), padding="SAME")
	#print(x.shape)
	return x

def convtrans_layer(layer, features, kernel_size):
  x = nn.ConvTranspose(features, kernel_size=kernel_size, strides=(1, 2, 2), padding="SAME")(layer)
  x = nn.relu(x)
  return x

class Encoder(nn.Module):
	@nn.compact
	def __call__(self, input_layer):
		x = conv_and_pool(input_layer, features=2, kernel_size=(3, 3, 1))
		x = conv_and_pool(x, features=4, kernel_size=(3, 3, 1))
		x = conv_and_pool(x, features=8, kernel_size=(3, 3, 1))
		x = conv_and_pool(x, features=16, kernel_size=(3, 3, 1))
		x = conv_and_pool(x, features=32, kernel_size=(3, 3, 1))
		#print("encoder")
		#print(x.shape) # == 68, 33, 2, 32

		return x
  
class Decoder(nn.Module):

	@nn.compact
	def __call__(self, input_layer):
		x = convtrans_layer(input_layer, 32, kernel_size=(3, 3, 1))
		x = convtrans_layer(x, 16, kernel_size=(3, 3, 1))
		x = convtrans_layer(x, 8, kernel_size=(3, 3, 1))
		x = convtrans_layer(x, 4, kernel_size=(3, 3, 1))
		x = convtrans_layer(x, 2, kernel_size=(3, 3, 1))
		x = nn.Conv(1, (3, 3, 1), padding="SAME")(x)
		#print("decoder")
		#print(x.shape)
		return x

class Autoencoder(nn.Module):
	def setup(self):
		self.encoder = Encoder()
		self.decoder = Decoder()

	def __call__(self, x):
		#print(args)
		#print(x)
		#print(kwargs)
		out = self.encoder(x)
		reconstructed = self.decoder(out)
		#print(f"out: {out}, rec: {reconstructed.shape}")
		return reconstructed, x 
	
	#@nn.module_method
	def generate(self, x):
		params = self.get_param("decoder")
		return Decoder.call(params, x)


#@jit
def compute_metrics(generated, real):
	error = generated - real
	loss = jnp.mean(jnp.square(error))
	metrics = {
		"loss": loss
	}		
	return metrics



class Dataset():
	def __init__(self):
		pass
	def __getitem__(self, i):
		return jnp.zeros((8, 2176, 1056, 2))

dataset = Dataset()

#@jit
def train_step(state):
	def loss_fn(params):
		data = dataset[0]
		#data = dataset[0][0]

		rec, out = Autoencoder().apply({"params": params}, data)
		#globals()["p"] = rec, out
		loss = compute_metrics(rec, out)["loss"]
		return jnp.array(loss)

	#optimizer, _, _ = optimizer.optimize(loss_fn)
	grads = jax.grad(loss_fn)(state.params)
	return state.apply_gradients(grads=grads)
	#return optimizer
  
#@jax.jit
#def eval_step(params, batch):
#  logits = CNN().apply({'params': params}, batch['image'])
#  return compute_metrics(logits, batch['label'])
print(f"{jax.device_count()} devices")

#_, autoencoder = Autoencoder.create_by_shape(KEY, [((2176, 1056, 2, 1), jnp.float32)])

optimizer = optim.sgd.GradientDescent(learning_rate=0.01)
EPOCHS = 10

state = train_state.TrainState.create(
	apply_fn=Autoencoder().apply,
	params=Autoencoder().init(rng, dataset[0])["params"],
	tx=optax.sgd(0.01)
)

for epoch in range(EPOCHS):
	bar = tqdm(range(100))
	for _ in bar:
		state = train_step(state)
		bar.set_description(f"Epoch {epoch}")

```
However, it doesn't work on the TPU and gives me the OOM error.

**Comment by nikhilanayak** [comment]
>  You should jit it for the code to compile on TPU
Will non-jitted code not run on the TPU?

**Comment by marcvanzee** [comment]
> > You should jit it for the code to compile on TPU
> 
> Will non-jitted code not run on the TPU?

Yes it does run it on the TPU, but you should always jit the code so it can be optimized by the XLA compiler. You should try to `jit` compile large code chunks (e.g., a train step function), so the XLA compiler is able to apply optimizations such as fusions so your code will run faster and consume less memory.

**Comment by nikhilanayak** [comment]
I jitted all the big code blocks and it runs fine, so my initial question is answered. However, when I do this, I also want to track the loss after each batch. The issue is that with jit, I can't coerce the Traced<ShapedArray>s to ints, so I can print them. Is there a way I can do this?

**Comment by marcvanzee** [comment]
Great, happy to hear it is resolved!

If you want to output the loss you can return it from the jitted function and output it then. This is done in all our example, I would recommend taking a look at the [mnist example](https://github.com/google/flax/blob/main/examples/mnist/train.py) and the [Annotated MNIST](https://flax.readthedocs.io/en/latest/notebooks/annotated_mnist.html) which contains more explanation.

**Comment by nikhilanayak** [comment]
I looked over the FLAX examples, but with my code:
```py
        def loss_fn(params):
                data = dataset[0]
                #data = dataset[0][0]

                rec, out = Autoencoder().apply({"params": params}, data)
                #globals()["p"] = rec, out
                loss = compute_metrics(rec, out)["loss"]
                return jnp.array(loss)

        grads = jax.grad(loss_fn)(state.params)
```
I never actually return the loss, so I don't have a way to get the loss itself. Am I using the wrong Jax/Flax functions?

**Comment by marcvanzee** [comment]
If you have taken a look at the examples, then you must have noticed that they don't use `jax.grad`, but `jax.grad_and_value`, which returns the gradient and the value. In your case you only return the gradient so you don't have the actual loss.

**Comment by nikhilanayak** [comment]
Thanks, that seems to get the loss **and** the grad.
Sorry to keep bothering you (especially since these questions aren't completely on topic) but this code
```py
state = train_state.TrainState.create(
           apply_fn=Autoencoder().apply,
           params=Autoencoder().init(rng, dataset[0])["params"],
           tx=optax.sgd(0.01)
    )
```
which was in the main code (not in a function), was causing an OOM, so I wrapped that in a jitted main. However, now I can't print out the loss because I'm still inside jitted code. I tried creating the state in a jit function and getting the value into a non-jit function, but that still causes OOMs. Am I going about this the wrong way or is there just no way to do this?
